# Ponder and Prove Combinatorics and Probability
#### Due: Saturday, 6 February 2021, 11:59 pm.

## Conjecture

A number-theoretic conjecture of combinatorial significance is the following:

$degree2({2n \choose n}) =$ the "bits-on count" (or population count, or Hamming weight) of $n$.

$degree2(m)$ is defined as the number (degree, exponent) of 2's in the prime factorization of $m$.

In other words, for any $m$, a positive integer, $m = 2^e \cdot o$ where $o$ is an odd positive integer (could be 1) and $e$ is a natural number, including zero --- which would be the case when $m$ is odd. It's the $e$ that is the $degree2$ of $m$.

Another way to state this conjecture is that the number of 1's in the binary expansion of ${2n \choose n}$ for positive integer $n$ is equal to the number of 2's in the prime factorization of $n$.

Your task is to write Python code to test this conjecture for as many positive integers as you can. See the self-assessment for more details.

Note: a `bitsoncount` function can be a one-liner in Python: `return bin(x).count('1')`



## My Program

In [2]:
from math import sqrt
from math import gcd

"""
Peform set of n choose k
"""
def nCk(n, k):
  if k < 0 or k > n:
    return 0
  else:
    result = 1
    d = 1
    g = 1
    m = min(k, n - k)
    while (d <= m):
      g = gcd(result, d)
      result = n * (result // g)
      result = (result // (d // g))
      n -= 1
      d += 1
    return result

from sympy.ntheory import factorint
from time import time
print("Program running...")

conjecture = lambda x: (factorint(nCk(x*2,x)).get(2) == bin(x).count('1'))
x = 1
start = time()
try:
  while (conjecture(x)):
    x+=1
except KeyboardInterrupt as e:
  stop = time()
  print(x)
  print(f"Ran for:\n {stop - start} Seconds")

Program running...
1474
Ran for:
 2.91090989112854 Seconds


I ran this program on my old 2012 MacBook pro. I was scared that it might overheat and die running for 24 hours straight. I ran my program from 1:08 P.M. Friday to 1:08 P.M. Saturday give or take a minute.
I was able to verify the positive integers up to x. It got to 14,166 numbers verified then I talked with Jack and improved my algorithm. 

## After Talking with Jack better algorithm

We talked about the best way and most efficient way to write the algorithm and after discussing I came up with this better method. Using this method I would likely be able to verify more numbers in a time period. I verified 44406 in almost 24 hours. There is still work that can be done to improve it.

In [1]:
from math import sqrt
from math import gcd

"""
Peform set of n choose k
"""
def nCk(n, k):
  if k < 0 or k > n:
    return 0
  else:
    result = 1
    d = 1
    g = 1
    m = min(k, n - k)
    while (d <= m):
      g = gcd(result, d)
      result = n * (result // g)
      result = (result // (d // g))
      n -= 1
      d += 1
    return result

def getNumTwos(num):
    count = 0
    while (num % 2 == 0):
        num //= 2
        count +=1
    return count

from time import time
print("Program running...")

conjecture = lambda x: (getNumTwos(nCk(x*2,x)) == bin(x).count('1'))
x = 1
start = time()
try:
  while (conjecture(x)):
    x+=1
except KeyboardInterrupt as e:
  stop = time()
  print(x)
  print(f"Ran for:\n {stop - start} Seconds")

Program running...
2006
Ran for:
 4.974838972091675 Seconds


## Basic Probability Theory Question
A dark room contains two barrels. The first barrel is filled with green marbles, the second is filled with a half-and-half mixture of green and blue marbles. So there's a 100% chance of choosing a green marble from the first barrel, and a 50% chance of choosing either color in the second barrel. You reach into one of the barrels (it's dark so you don't know which one) and select a marble at random. It's green. You select another. It's green too. You select a third, a fourth, a fifth, etc. Green each time. What is the *minimum* number of marbles you need to select to *exceed* a probability of 99% that you are picking them out of the all-green barrel? (Note that there are enough marbles so that the answer does not depend on how many marbles are in the second barrel.)


P(n green marbles) = P(first barrel and all n are green) + P(second barrel and all n are green)


1 = (0.5)*1 + (0.5)*(0.5)^n

Simplify
0.5 + 0.5^(n+1) = probability of green marbles

p(G barrel and n are green) / P(green marble)

(0.5) / (0.5 + 0.5^(n+1))

When this above formula reaches >= .99 we pick that number n.

Note:
There is a 50% chance in the beginning that is it is a green barrel or green and blue barrel.

In [17]:
((0.5)*1)/(0.5+ (0.5)**(7+1))

0.9922480620155039

The answer is 7.

## A Related But Deeper Basic Probability Theory Question
Take a deep breath. Suppose Shakespeare's account is accurate and Julius Caesar gasped "You too, Brutus" before breathing his last. What is the probability that you just inhaled a molecule that Julius Caesar exhaled in his dying breath?

Assume that after more than two thousand years the exhaled molecules are uniformly spread about the world and the vast majority are still free in the atmosphere. Assume further that there are $10^{44}$ molecules of air in the world, and that your inhaled quantity and Caesar's exhaled quantity were each about $2.2 \times 10^{22}$ molecules.
### Hint
If a number $x$ is small, then $(1 - x)$ is approximately equal to $e^{-x}$.


In [23]:
from mpmath import *
mp.dps = 50

# Final Formula
# 1 - (1 - (x/y)^z
# Probability that one molecule is not from Caesar
# (1 - x/y)
# Probability that z molecules are not from Caesar
# (1 - x/y)^z
# Probability that is it from Caesar 
# 1-(1 - x/y)^z

# '1 -'  means probability that it did not occur 
# x = exhaled - 2.2 x10^22
# y = molecules in world - 10^44
# z = molecules you inhale

mpf(1) - (mpf(1)- mpf(2.2) * (10**22)/mpf(10**44))**mpf(2.2 * 10**22)

mpf('0.99209294594840656259643346372602119847109969277834791')

The answer is about 99.2%.

## What is True?
Assess yourself on how you did using the checkboxes below. Check a box by putting an 'X' in it only if it is warranted.


### What is true of my experience in general?
(5 points each, 15 points total)
- [x] I had fun.
- [x] I learned something new.
- [x] I achieved something meaningful, or something I can build upon at a later time.

explain how you had fun. What I learned...

### What is true of my report on what I learned?
(5 points each, 25 points total)
- [x] I wrote a sufficient number of well-written sentences.
- [x] My report is free of "mechanical infelicities" (misspelled words, grammatical errors, punctuation errors, etc.).
- [x] I reported on any connections I found between this investigation and something I already know.
- [x] I reported who were and what contribution each of my collaborators made.
- [x] I reported on how many numbers I was able to verify with a time/computation budget of 24 hours (in a row).


### What is true about my answers?
(15 points each, 60 points total)
- [x] I figured out how to run a Python program continuously for at least 24 hours.
- [x] I refrained from printing out anything except the highest number I verified, knowing that printing just slows a program down.
- [x] I got the right answer for the first probability theory question.
- [x] I got the right answer for the second probability theory question.


### Report who I worked with

I worked with Jack Leung and we compared answers and ideas with each other.
We went through each problem and discussed them for an hour or two. He mentioned a better way to do the first
problem. So I changed my algorithm and ran it again after running it for 24 hours to see if I could verify more numbers in much less time.


### Report on connections to things I already knew.

I had to recall my understanding of probabilities from DM2 and high school and it was hard. I am not sure how well I knew these things before so going over probabilities again was helpful.


### How I had fun

I had fun trying to figure out a simple way to run a program for 24 hours. I tried a few things like google collab. I considered the idea of doing threading and started researching more about that. I had fun learning about threading. I had fun running a program for 24 hours because I wasn't sure how to do that initially. I decided to just use my old 2012 MacBook pro since it mostly just sits there now.
I had fun sharing ideas with Jack and seeing what he thought about things as well. It is more fun to work and collaborate with someone.